README.md
# S3LevelDown

An [abstract-leveldown](https://github.com/Level/abstract-leveldown) compliant implementation of [LevelDOWN](https://github.com/Level/leveldown) that uses [Amazon S3](https://aws.amazon.com/s3/) as a backing store. S3 is actually a giant key-value store on the cloud, even though it is marketed as a file store. Use this database with the [LevelUP](https://github.com/Level/levelup/) API.

To use this optimally, please read "Performance considerations" and "Warning about concurrency" sections below.

You could also use this as an alternative API to read/write S3. The API simpler to use when compared to the AWS SDK!

## Installation

Install `s3leveldown` and peer dependencies `levelup` and `@aws-sdk/client-s3` with `yarn` or `npm`.

```bash
$ npm install s3leveldown @aws-sdk/client-s3 levelup
```

## Documentation

See the [LevelUP API](https://github.com/Level/levelup#api) for high level usage.

### `new S3LevelDown(location [, s3])`

Constructor of `s3leveldown` backing store. Use with `levelup`.

Arguments:
* `location` name of the S3 bucket with optional sub-folder. Example `mybucket` or `mybucket/folder`.
* `s3` Optional `S3` client from `aws-sdk`. A default client will be used if not specified.

## Example

Please refer to the [AWS SDK docs to set up your API credentials](http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html) before using.

### Using Promises

```js
const levelup = require('levelup');
const S3LevelDown = require('s3leveldown');

(async () => {
  // create DB
  const db = levelup(new S3LevelDown('mybucket'));

  // put items
  await db.batch()
    .put('name', 'Pikachu')
    .put('dob', 'February 27, 1996')
    .put('occupation', 'Pokemon')
    .write();
  
  // read items
  await db.createReadStream()
    .on('data', data => { console.log('data', `${data.key.toString()}=${data.value.toString()}`); })
    .on('close', () => { console.log('done!') });
})();
```

### Using Callbacks

```js
const levelup = require('levelup');
const S3LevelDown = require('s3leveldown');

const db = levelup(new S3LevelDown('my_bucket'));

db.batch()
  .put('name', 'Pikachu')
  .put('dob', 'February 27, 1996')
  .put('occupation', 'Pokemon')
  .write(function () { 
    db.readStream()
      .on('data', console.log)
      .on('close', function () { console.log('Pika pi!') })
  });
```

### Example with min.io

You could also use s3leveldown with S3 compatible servers such as [MinIO](https://github.com/minio/minio).

```js
const levelup = require('levelup');
const S3LevelDown = require('s3leveldown');
const AWS = require('aws-sdk');

const s3 = new AWS.S3({
  apiVersion: '2006-03-01',
  accessKeyId: 'YOUR-ACCESSKEYID',
  secretAccessKey: 'YOUR-SECRETACCESSKEY',
  endpoint: 'http://127.0.0.1:9000',
  s3ForcePathStyle: true,
  signatureVersion: 'v4'
});

const db = levelup(new S3LevelDown('my_bucket', s3));
```

### [Example with PouchDB](./examples/pouchdb)

## Sub folders

You can create your Level DB in a sub-folder in your S3 bucket, just use `my_bucket/sub_folder` when passing the location.

## Performance considerations

There are a few performance caveats due to the limited API provided by the AWS S3 API:

* When iterating, getting values is expensive. A seperate S3 API call is made to get the value of each key. If you don't need the value, pass `{ values: false }` in the options. Each S3 API call can return 1000 keys, so if there are 3000 results, 3 calls are made to list the keys, and if getting values as well, another 3000 API calls are made.

* Avoid iterating large datasets when passing `{ reverse: true }`. Since the S3 API call do not allow retrieving keys in reverse order, the entire result set needs to be stored in memory and reversed. If your database is large ( >5k keys ), be sure to provide start (`gt`, `gte`) and end (`lt`, `lte`), or the entire database will need to be fetched.

* By default when iterating, 1000 keys will be returned. If you only want 10 keys for example, set `{ limit: 10 }` and the S3 API call will only request 10 keys. Note that if you have `{ reverse: true }`, this optimisation does not apply as we need to fetch everything from start to end and reverse it in memory. To override the default number of keys to return in a single API call,  you can set the ` s3ListObjectMaxKeys` option when creating the iterator. The maximum accepted by the S3 API is 1000.

* Specify the AWS region of the bucket to improve performance, by calling `AWS.config.update({ region: 'ap-southeast-2' });` replace `ap-southeast-2` with your region.

## Warning about concurrency

Individual operations (`put` `get` `del`) are atomic as guaranteed by S3, but the implementation of `batch` is not atomic. Two concurrent batch calls will have their operations interwoven. Don't use any plugins which require this to be atomic or you will end up with your database corrupted! However, if you can guarantee that only one process will write the S3 bucket at a time, then this should not be an issue. Ideally, you want to avoid race conditions where two processes are writing to the same key at the same time. In those cases the last write wins.

Iterator snapshots are not supported. When iterating through a list of keys and values, you may get the changes, similar to dirty reads.

## Tests and debug

S3LevelDown uses [debug](https://github.com/visionmedia/debug). To see debug message set the environment variable `DEBUG=S3LevelDown`.

To run the test suite, you need to set a S3 bucket to the environment variable `S3_TEST_BUCKET`. Also be sure to [set your AWS credentials](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html)

```bash
$ S3_TEST_BUCKET=my-test-bucket npm run test
```

## License

MIT


CHANGELOG.md
# Changelog

## [3.0.0] - 2023-04-05

### Changed

- When initialising S3LevelDown, new is always required.
- Update dependencies
- Node.js 14 or higher required.
- `levelup` `^5.1.1` required.
- `@aws-sdk/client-s3` (AWS SDK v3) required. (replaces `aws-sdk`)
- Use ESNext syntax.

## [2.2.2] - 2021-11-07

### Changed

- Move `levelup` to peer dependency and support `^5.1.1`.

## [2.2.1] - 2021-02-14

### Changed

- Improve TypeScript definition.

## [2.2.0] - 2021-02-05

### Added

- TypeScript types.

### Changed

- Update dependencies.

## [2.1.0] - 2020-04-05

### Changed

- Update `abstract-leveldown`.

### Fixed

- S3 client now passed to constructor in function (#2).

## [2.0.0] - 2019-04-27

### Changed

- Support LevelDOWN `4.0.1`.
- Support new `abstract-leveldown` `6.0.3` tests.
- Constructor now takes `location` and `s3` parameters.
- `open` now creates the S3 bucket if `createIfMissing` is `true`

## [1.0.0] - 2017-02-05

- Initial Release


test.js
var test = require('tape')
var suite = require('abstract-leveldown/test')
var S3LevelDown = require('./s3leveldown')

if (!process.env.S3_TEST_BUCKET) {
  console.error("Please set the S3_TEST_BUCKET environment variable to run the test")
  process.exit(1)
  return
}

var prefix = "/__leveldown_test-" + Date.now();

var bucketTestIndex = 0;
var testCommon = suite.common({
  test: test,
  factory: function () {
    return new S3LevelDown(process.env.S3_TEST_BUCKET + prefix + "-" + (++bucketTestIndex))
  },
  snapshots: false,
  seek: false,
  bufferKeys: false,
  createIfMissing: false,
  errorIfExists: false
})

suite(testCommon)

// custom tests

var db

test('setUp common', testCommon.setUp)

test('setUp #1', function (t) {
  db = testCommon.factory()
  db.open(function () {
    db.batch()
      .put('foo', 'bar')
      .put('foo2', 'bar2')
      .put('foo3', 'bar3')
      .write(function() { t.end(); })
  })
})

test('lazy iterator next delete next', function (t) {
  var iterator = db.iterator()
  iterator.next(function (err, key, value) {
    t.equal(err, null, 'no errors')
    t.equal(key.toString(), 'foo', 'correct key')
    db.del('foo2', function () {
      iterator.next(function (err, key, value) {
        t.error(err)
        t.equal(key.toString(), "foo3", 'correct key')
        iterator.end(t.end.bind(t))
      })
    })
  })
})


testCommon.js
var path      = require('path')
  , leveldown = require('./s3leveldown')

var dbidx = 0
  , bucket = process.env.S3_TEST_BUCKET
  , location = function () {
      return bucket + "/__leveldown_test-" + (++dbidx)
    }

  , lastLocation = function () {
      return bucket + "/__leveldown_test-" + dbidx
    }

  , cleanup = function (callback) {
      var db = leveldown(lastLocation())
      db.open(function () {
        collectEntries(db.iterator({ keyAsBuffer: false, values: false }), function (err, data) {
          for (var i in data) {
            data[i].type = 'del'
          }

          db.batch(data, callback)
        })
      })
      
    }

  , setUp = function (t) {
      cleanup(function (err) {
        t.error(err, 'cleanup returned an error')
        t.end()
      })
    }

  , tearDown = function (t) {
      setUp(t) // same cleanup!
    }

  , collectEntries = function (iterator, callback) {
      var data = []
        , next = function () {
            iterator.next(function (err, key, value) {
              if (err) return callback(err)
              if (!arguments.length) {
                return iterator.end(function (err) {
                  callback(err, data)
                })
              }
              data.push({ key: key, value: value })
              setTimeout(next, 0)
            })
          }
      next()
    }

module.exports = {
    location       : location
  , cleanup        : cleanup
  , lastLocation   : lastLocation
  , setUp          : setUp
  , tearDown       : tearDown
  , collectEntries : collectEntries
}


s3leveldown.js
const AbstractLevelDOWN = require('abstract-leveldown').AbstractLevelDOWN
  , AbstractIterator = require('abstract-leveldown').AbstractIterator
  , ltgt = require('ltgt')
  , debug = require('debug')('S3LevelDown')
  , AWS = require('@aws-sdk/client-s3')

const staticS3 = new AWS.S3Client({ apiVersion: '2006-03-01' })

function lt(value) {
  return ltgt.compare(value, this._finish) < 0
}

function lte(value) {
  return ltgt.compare(value, this._finish) <= 0
}

function getStartAfterKey(key) {
  const keyMinusOneNum = (key.charCodeAt(key.length - 1) - 1)
  const keyMinusOne = keyMinusOneNum >= 0 ? (String.fromCharCode(keyMinusOneNum) + '\uFFFF') : ''
  return key.substring(0, key.length - 1) + keyMinusOne
}

function nullEmptyUndefined(v) {
  return typeof v === 'undefined' || v === null || v === ''
}

class S3Iterator extends AbstractIterator {
  constructor(db, options) {
    super(db)
    const self = this
    self._limit = options.limit

    if (self._limit === -1)
      self._limit = Infinity

    self.keyAsBuffer = options.keyAsBuffer !== false
    self.valueAsBuffer = options.valueAsBuffer !== false
    self.fetchValues = options.values
    self._reverse = options.reverse
    self._options = options
    self._done = 0
    self.bucket = db.bucket
    self.db = db
    self.s3ListObjectMaxKeys = options.s3ListObjectMaxKeys || 1000
    if (!self._reverse && self._limit < self.s3ListObjectMaxKeys) {
      self.s3ListObjectMaxKeys = self._limit
    }

    self._start = ltgt.lowerBound(options)
    self._finish = ltgt.upperBound(options)
    if (!nullEmptyUndefined(self._finish)) {
      if (ltgt.upperBoundInclusive(options))
        self._test = lte

      else
        self._test = lt
    }

    if (!nullEmptyUndefined(self._start))
      self.startAfter = ltgt.lowerBoundInclusive(options) ? getStartAfterKey(self._start) : self._start

    debug('new iterator %o', self._options)
  }

  _next(callback) {
    const self = this

    if (self._done++ >= self._limit ||
      (self.data && self.dataUpto == self.data.length && !self.s3nextContinuationToken))
      return setImmediate(callback)

    if (!self.data || self.dataUpto == self.data.length) {
      listObjects()
    } else {
      fireCallback()
    }

    function listObjects() {
      const params = {
        Bucket: self.bucket,
        MaxKeys: self.s3ListObjectMaxKeys
      }

      if (self.db.folderPrefix !== '') {
        params.Prefix = self.db.folderPrefix
      }

      if (self.s3nextContinuationToken) {
        params.ContinuationToken = self.s3nextContinuationToken
        debug('listObjectsV2 ContinuationToken %s', params.ContinuationToken)
      }
      else if (typeof self.startAfter !== 'undefined') {
        params.StartAfter = self.db.folderPrefix + self.startAfter
      }

      self.db.s3.send(new AWS.ListObjectsV2Command(params), function (err, data) {
        if (err) {
          debug('listObjectsV2 error %s', err.message)
          callback(err)
        } else {
          if (data.KeyCount === 0) {
            debug('listObjectsV2 empty')
            return setImmediate(callback)
          }

          debug('listObjectsV2 %d keys', data.KeyCount)

          if (self.data && self.dataUpto === 0) {
            self.data = self.data.concat(data.Contents)
          } else {
            self.data = data.Contents
          }

          self.dataUpto = 0
          self.s3nextContinuationToken = data.NextContinuationToken

          if (self._reverse && self.s3nextContinuationToken &&
            data.Contents.every(function (x) {
              return self._test(x.Key.substring(self.db.folderPrefix.length, x.Key.length))
            })) {
            listObjects()
          } else {
            fireCallback()
          }
        }
      })
    }

    function fireCallback() {
      let index, key
      for (; ;) {
        index = (!self._reverse) ? self.dataUpto : (self.data.length - 1 - self.dataUpto)
        const awskey = self.data[index].Key
        key = awskey.substring(self.db.folderPrefix.length, awskey.length)
        debug('iterator data index %d: %s', index, key)
        self.dataUpto++

        if (self._test(key)) {
          break
        }

        if (!self._reverse || self.dataUpto === self.data.length) {
          return setImmediate(callback)
        }
      }

      if (self.fetchValues) {
        if (self.data[index].Size === 0)
          getCallback(null, '')
        else
          self.db._get(key, { asBuffer: self.valueAsBuffer }, getCallback)
      }
      else
        getCallback()

      function getCallback(err, value) {
        debug('iterator data getCallback %s = %s', key, value)
        if (err) {
          if (err.message == 'NotFound') {
            // collection changed while we were iterating, skip this key
            return setImmediate(function () {
              self._next(callback)
            })
          }
          return setImmediate(function () {
            callback(err)
          })
        }

        if (self.keyAsBuffer && !(key instanceof Buffer))
          key = Buffer.from(key)
        if (!self.keyAsBuffer && (key instanceof Buffer))
          key = key.toString('utf8')

        if (self.fetchValues) {
          if (self.valueAsBuffer && !(value instanceof Buffer))
            value = Buffer.from(value)
          if (!self.valueAsBuffer && (value instanceof Buffer))
            value = value.toString('utf8')
        }

        setImmediate(function () {
          debug('_next result %s=%s', key, value)
          callback(null, key, value)
        })
      }
    }
  }

  _test() { return true }
}


class S3LevelDown extends AbstractLevelDOWN {
  constructor(location, s3) {
    super()
    if (typeof location !== 'string') {
      throw new Error('constructor requires a location string argument')
    }

    this.s3 = s3 || staticS3

    if (location.indexOf('/') !== -1) {
      this.folderPrefix = location.substring(location.indexOf('/') + 1, location.length) + '/'
      this.bucket = location.substring(0, location.indexOf('/'))
    } else {
      this.folderPrefix = ''
      this.bucket = location
    }

    debug('db init %s %s', this.bucket, this.folderPrefix)
  }

  _open(options, callback) {
    this.s3.send(new AWS.HeadBucketCommand({ Bucket: this.bucket }), (err) => {
      if (err) {
        // error, bucket is not found
        if (options.createIfMissing && err['$metadata'].httpStatusCode === 404) {
          // try to create it
          this.s3.send(new AWS.CreateBucketCommand({ Bucket: this.bucket }), (err) => {
            if (err) {
              setImmediate(() => callback(err))
            } else {
              setImmediate(callback)
            }
          })
        } else {
          setImmediate(() => callback(new Error(`Bucket ${this.bucket} does not exists or is inaccessible`)))
        }
      } else {
        setImmediate(callback)
      }
    })
  }

  _put(key, value, options, callback) {
    if (nullEmptyUndefined(value))
      value = Buffer.from('')

    if (!(value instanceof Buffer || value instanceof String))
      value = String(value)

    this.s3.send(new AWS.PutObjectCommand({
      Bucket: this.bucket,
      Key: this.folderPrefix + key,
      Body: value
    }), function (err) {
      if (err) {
        debug('Error s3 upload: %s %s', key, err.message)
        callback(err)
      } else {
        debug('Successful s3 upload: %s', key)
        callback()
      }
    })
  }

  _get(key, options, callback) {
    this.s3.send(new AWS.GetObjectCommand({
      Bucket: this.bucket,
      Key: this.folderPrefix + key
    }), async function (err, data) {
      if (err) {
        debug('Error s3 getObject: %s %s', key, err.message)
        if (err.Code === 'NoSuchKey') {
          callback(new Error('NotFound'))
        } else {
          callback(err)
        }
      } else {
        let value
        try {
          debug('s3 getObject callback as %s: %s', options.asBuffer ? 'buf' : 'string', key)
          if (options && options.asBuffer) {
            const byteArray = await data.Body?.transformToByteArray()
            value = Buffer.from(byteArray.buffer, byteArray.byteOffset, byteArray.byteLength)
          } else {
            value = await data.Body?.transformToString('utf8')
          }
        } catch (err) {
          callback(err, null)
          return
        }
        callback(null, value)
      }
    })
  }

  _del(key, options, callback) {
    this.s3.send(new AWS.DeleteObjectCommand({
      Bucket: this.bucket,
      Key: this.folderPrefix + key
    }), function (err) {
      if (err) {
        debug('Error s3 delete: %s %s', key, err.message)
        callback(err)
      } else {
        debug('Successful s3 delete: %s', key)
        callback()
      }
    })
  }

  _batch(array, options, callback) {
    const len = array.length, self = this
    let i = 0;

    function act(action, cb) {
      if (!action) {
        return setImmediate(cb)
      }

      const key = (action.key instanceof Buffer) ? action.key : String(action.key)
      const value = action.value

      if (action.type === 'put') {
        self._put(key, value, null, cb)
      } else if (action.type === 'del') {
        self._del(key, null, cb)
      }
    }

    function actCallback(err) {
      if (err) {
        return setImmediate(function () { callback(err) })
      }

      if (++i >= len) {
        return setImmediate(callback)
      }

      act(array[i], actCallback)
    }

    act(array[i], actCallback)
  }

  _iterator(options) {
    return new S3Iterator(this, options)
  }
}

module.exports = S3LevelDown
